### PR TITLE
Add Ability to Tweak JWT Refresh Intervals

### PR DESCRIPTION
**Pull request by say25**
Summary of the changes (Less than 80 chars)
 - Add Ability to Tweak JWT Refresh Intervals

### CODE REVIEW

**Comment by Tratcher** [comment]
- The release/3.0 branch is not open, please retarget to master.
- Adding new APIs requires [updating](https://github.com/aspnet/AspNetCore/blob/master/docs/ReferenceAssemblies.md) reference assemblies, but don't do that yet, I expect we'll have feedback about the names.

**Comment by Tratcher** [comment]
Is this intended to help address https://github.com/aspnet/AspNetCore/issues/13046? Or some other active issue?

**Comment by Tratcher** [comment]
If we're going to do this it should also be done for OpenIdConnect.
https://github.com/aspnet/AspNetCore/blob/bcc962a182a273cdaa7de474528e97913373a212/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectOptions.cs#L132

File: src/Security/Authentication/JwtBearer/src/JwtBearerOptions.cs [hunk 1]
```diff
@@ -111,5 +111,15 @@ public new JwtBearerEvents Events
...
+
+        /// <summary>
+        /// 1 day is the default time interval that afterwards, <see cref="ConfigurationManager" /> will obtain new configuration.
+        /// </summary>
+        public TimeSpan AutomaticRefreshInterval { get; set; } = ConfigurationManager<OpenIdConnectConfiguration>.DefaultAutomaticRefreshInterval;
+
+        /// <summary>
+        /// 30 seconds is the default time interval that must pass for <see cref="ConfigurationManager" /> to obtain a new configuration.
+        /// </summary>
+        public TimeSpan RefreshInterval { get; set; } = ConfigurationManager<OpenIdConnectConfiguration>.DefaultRefreshInterval;
```
Comment (Tratcher) [inline src/Security/Authentication/JwtBearer/src/JwtBearerOptions.cs]:
```suggestion
        public TimeSpan MinimumRefreshInterval { get; set; } = ConfigurationManager<OpenIdConnectConfiguration>.DefaultRefreshInterval;
```

File: src/Security/Authentication/JwtBearer/src/JwtBearerOptions.cs [hunk 2]
```diff
@@ -111,5 +111,15 @@ public new JwtBearerEvents Events
...
         /// </summary>
         public bool IncludeErrorDetails { get; set; } = true;
+
+        /// <summary>
+        /// 1 day is the default time interval that afterwards, <see cref="ConfigurationManager" /> will obtain new configuration.
+        /// </summary>
+        public TimeSpan AutomaticRefreshInterval { get; set; } = ConfigurationManager<OpenIdConnectConfiguration>.DefaultAutomaticRefreshInterval;
+
+        /// <summary>
+        /// 30 seconds is the default time interval that must pass for <see cref="ConfigurationManager" /> to obtain a new configuration.
```
Comment (Tratcher) [inline src/Security/Authentication/JwtBearer/src/JwtBearerOptions.cs]:
```suggestion
        /// 30 seconds is the default, minimum time interval that must pass for <see cref="ConfigurationManager" /> to obtain a new configuration.
```

**Comment by say25** [comment]
> If we're going to do this it should also be done for OpenIdConnect.
> https://github.com/aspnet/AspNetCore/blob/bcc962a182a273cdaa7de474528e97913373a212/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectOptions.cs#L132

Done

> * The release/3.0 branch is not open, please retarget to master.
> * Adding new APIs requires [updating](https://github.com/aspnet/AspNetCore/blob/master/docs/ReferenceAssemblies.md) reference assemblies, but don't do that yet, I expect we'll have feedback about the names.

Done

> Is this intended to help address #13046? Or some other active issue?

Partially though I'm basically trying to use the simple `services.AddAuthentication().AddJwtBearer` method and tweak these values

File: src/Security/Authentication/JwtBearer/src/JwtBearerOptions.cs (hunk 1 above)
Comment (say25) [reply src/Security/Authentication/JwtBearer/src/JwtBearerOptions.cs]: Not sure this makes sense as we are settings `RefreshInterval` not `MinimumRefreshInterval` below

File: src/Security/Authentication/JwtBearer/src/JwtBearerOptions.cs (hunk 2 above)
Comment (Tratcher) [reply src/Security/Authentication/JwtBearer/src/JwtBearerOptions.cs]: Even if we don't rename the property, this clarification in the docs is still warranted.

**Review by Tratcher — COMMENTED** [review]
Go ahead and update the ref assemblies please. https://github.com/aspnet/AspNetCore/blob/master/docs/ReferenceAssemblies.md

**Comment by say25** [comment]
@Tratcher Im running into some build issues on `master`. From what I can tell I should be good but this is my first contribution to this project so maybe I missed something?

```PS D:\work\Github\AspNetCore> dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   5.0.100-alpha1-014696
 Commit:    e1f85b6362

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17763
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   D:\work\Github\AspNetCore\.dotnet\sdk\5.0.100-alpha1-014696\

Host (useful for support):
  Version: 5.0.0-alpha1.19467.5
  Commit:  708eb333f0

.NET Core SDKs installed:
  5.0.100-alpha1-014696 [D:\work\Github\AspNetCore\.dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.App 5.0.0-alpha1.19466.7 [D:\work\Github\AspNetCore\.dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 5.0.0-alpha1.19467.5 [D:\work\Github\AspNetCore\.dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 5.0.0-alpha1.19467.5 [D:\work\Github\AspNetCore\.dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```

**Comment by Tratcher** [comment]
What kind of build issues? The "Build Code Check" build is asking you to update the reference assemblies.

**Comment by say25** [comment]
When I try to run `Regenerate reference assemblies for all projects
Run .\eng\scripts\GenerateReferenceAssemblies.ps1 from repository root.`

I get a bunch of errors like: 
```
D:\work\Github\AspNetCore\src\ProjectTemplates\Web.Spa.ProjectTemplates\Microsoft.DotNet.Web.Spa.ProjectTemplates.cspro
j : error MSB4236: The SDK 'Microsoft.NET.Sdk' specified could not be found.
D:\work\Github\AspNetCore\src\ProjectTemplates\testassets\DotNetToolsInstaller\DotNetToolsInstaller.csproj(4,31): error
 MSB4236: The SDK 'Microsoft.NET.Sdk' specified could not be found.
```

**Comment by Tratcher** [comment]
Make sure you run restore.cmd first.

**Comment by say25** [comment]
```PS D:\work\Github\AspNetCore> .\restore.cmd
Building of C# project is enabled and has dependencies on NodeJS projects. Building of NodeJS projects is enabled since node is detected in C:\Program Files.
Detected JDK in C:\Program Files\Java\jdk1.8.0_211 (via JAVA_HOME)
  It was not possible to find any installed .NET Core SDKs
  Did you mean to run .NET Core SDK commands? Install a .NET Core SDK from:
      https://aka.ms/dotnet-download
C:\Users\syeh\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19462.4\tools\Tools.proj : error : Unable to locat
e the .NET Core SDK. Check that it is installed and that the version specified in global.json (if any) matches the inst
alled version.
C:\Users\syeh\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19462.4\tools\Tools.proj : error MSB4236: The SDK
'Microsoft.NET.Sdk' specified could not be found.
Build failed.
```

**Comment by Tratcher** [comment]
Hmm, thanks for trying. It's working locally for me so I updated the refs and pushed to your branch.

**Comment by say25** [comment]
Thanks. Out of curiosity. Will this get backported to 3.X?

**Comment by Tratcher** [comment]
> Thanks. Out of curiosity. Will this get backported to 3.X?

Unlikely, this doesn't unblock any critical scenarios, it just make existing scenarios slightly easier.

**Comment by Tratcher** [comment]
Found the build issue:
https://github.com/aspnet/AspNetCore/blob/master/docs/BuildFromSource.md#common-error-unable-to-locate-the-net-core-sdk

**Comment by Tratcher** [comment]
@blowdart for triage today.

**Comment by anurse** [comment]
@aspnet/asp-net-api-reviews @davidfowl @rynowak API review pls

**Review by Tratcher — APPROVED** [review]

File: src/Security/Authentication/OpenIdConnect/src/OpenIdConnectOptions.cs [hunk 3]
```diff
@@ -327,5 +327,15 @@ public override CookieOptions Build(HttpContext context, DateTimeOffset expiresF
...
             }
         }
+
+        /// <summary>
+        /// 1 day is the default time interval that afterwards, <see cref="ConfigurationManager" /> will obtain new configuration.
+        /// </summary>
+        public TimeSpan AutomaticRefreshInterval { get; set; } = ConfigurationManager<OpenIdConnectConfiguration>.DefaultAutomaticRefreshInterval;
+
+        /// <summary>
+        /// 30 seconds is the default time interval that must pass for <see cref="ConfigurationManager" /> to obtain a new configuration.
```
Comment (say25) [inline src/Security/Authentication/OpenIdConnect/src/OpenIdConnectOptions.cs]: For consistency with `JwtBearerOptions.cs` changes
```suggestion
        /// 30 seconds is the default, minimum time interval that must pass for <see cref="ConfigurationManager" /> to obtain a new configuration.
```

**Comment by Tratcher** [comment]
*ping* @aspnet/asp-net-api-reviews @davidfowl @rynowak API review pls

File: src/Security/Authentication/JwtBearer/src/JwtBearerOptions.cs [hunk 4]
```diff
@@ -111,5 +111,15 @@ public new JwtBearerEvents Events
...
         /// </summary>
         public bool IncludeErrorDetails { get; set; } = true;
+
+        /// <summary>
+        /// 1 day is the default time interval that afterwards, <see cref="ConfigurationManager" /> will obtain new configuration.
+        /// </summary>
+        public TimeSpan AutomaticRefreshInterval { get; set; } = ConfigurationManager<OpenIdConnectConfiguration>.DefaultAutomaticRefreshInterval;
+
+        /// <summary>
+        /// 30 seconds is the default, minimum time interval that must pass for <see cref="ConfigurationManager" /> to obtain a new configuration.
```
Comment (halter73) [inline src/Security/Authentication/JwtBearer/src/JwtBearerOptions.cs]: The [doc comments for the ConfigurationManager](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/06cae9a6d422dac7b76f1db9de032b98c1d8c836/src/Microsoft.IdentityModel.Protocols/Configuration/ConfigurationManager.cs#L131-L159) make it clearer why AutomaticRefreshInterval and RefreshInterval exist as two different properties. I would just change these doc comments to match. E.g.

```suggestion
        /// The minimum time between <see cref="ConfigurationManager" /> retrievals, in the event that a retrieval failed, or that a refresh was explicitly requested. 30 seconds is the default.
```

I don't care about the exact wording, but including "in the event that a retrieval failed, or that a refresh was explicitly requested" makes it a lot clearer to me.

File: src/Security/Authentication/OpenIdConnect/src/OpenIdConnectOptions.cs [hunk 5]
```diff
@@ -327,5 +327,15 @@ public override CookieOptions Build(HttpContext context, DateTimeOffset expiresF
...
             }
         }
+
+        /// <summary>
+        /// 1 day is the default time interval that afterwards, <see cref="ConfigurationManager" /> will obtain new configuration.
+        /// </summary>
+        public TimeSpan AutomaticRefreshInterval { get; set; } = ConfigurationManager<OpenIdConnectConfiguration>.DefaultAutomaticRefreshInterval;
+
+        /// <summary>
+        /// 30 seconds is the default, minimum time interval that must pass for <see cref="ConfigurationManager" /> to obtain a new configuration.
```
Comment (say25) [inline src/Security/Authentication/OpenIdConnect/src/OpenIdConnectOptions.cs]: For consistency:
```suggestion
        /// The minimum time between <see cref="ConfigurationManager" /> retrievals, in the event that a retrieval failed, or that a refresh was explicitly requested. 30 seconds is the default.
```

**Review by halter73 — APPROVED** [review]

**Comment by Tratcher** [comment]
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=400582

**Comment by aspnet-hello** [comment]
This comment was made automatically. If there is a problem contact aspnetcore-build@microsoft.com.

I've triaged the above build. I've created/commented on the following issue(s)
https://github.com/aspnet/AspNetCore-Internal/issues/3297

**Comment by Tratcher** [comment]
Thanks

**Pull request merged**